### ISSUE TITLE
Plugin submission error group 2: sanitize, escape, validate

### ISSUE THREAD

**Issue by jonathanstegall** [MinnPost/object-sync-for-salesforce#87]
As in #86, this is the email from the initial WordPress plugin review.

> There are issues with your plugin code.
> 
> Please read this ENTIRE email, address all listed issues, and reply to this email with your corrected code attached. It is required for you to read and reply to these emails, and failure to do so will result in significant delays with your plugin being accepted.
> 
> Also please remember in addition to code quality, security and functionality, we require all plugins adhere to our guidelines. If you have not yet, please read them:
> 
> * https://developer.wordpress.org/plugins/wordpress-org/detailed-plugin-guidelines/

## Please sanitize, escape, and validate your POST calls

When you include POST/GET/REQUEST calls in your plugin, it's important to sanitize, validate, and escape them. The goal here is to prevent a user from accidentally sending trash data through the system, as well as protecting them from potential security issues.

SANITIZE: All instances where generated content is inserted into the database, or into a file, or being otherwise processed by WordPress, the data MUST be properly sanitized for security. By sanitizing your POST data when used to make action calls or URL redirects, you will lessen the possibility of XSS vulnerabilities. You should never have a raw data inserted into the database, even by a update function, and even with a prepare() call.

VALIDATE: In addition to sanitization, you should validate all your calls. If a $_POST call should only be a number, ensure it's an int() before you pass it through anything. Even if you're sanitizing or using WordPress functions to ensure things are safe, we ask you please validate for sanity's sake. Any time you are adding data to the database, it should be the right data.

ESCAPE: Similarly, when you're outputting data, make sure to escape it properly, so it can't hijack admin screens. There are many esc_*() functions you can use to make sure you don't show people the wrong data.

In all cases, using stripslashes or strip_tags is not enough. You need to use the most appropriate method associated with the type of content you're processing. Check that a URL is a URL and don't just be lazy and use sanitize_text please. The ultimate goal is that you should ensure that invalid and unsafe data is NEVER processed or displayed. Clean everything, check everything, escape everything, and never trust the users to always have input sane data.

Please review this document and update your code accordingly: http://codex.wordpress.org/Validating_Sanitizing_and_Escaping_User_Data

Example:

```
set_transient( $cachekey, $_POST, 0 );
$wordpress_fields = $this->get_wordpress_object_fields( $_POST['wordpress_object'] );
```

----

> Please make sure you've addressed ALL issues brought up in this email. When you've corrected your code, reply to this email with the updated code attached as a zip, or provide a link to the new code for us to review. If you use gmail, you won’t be able to send a ZIP file if it contains any JS files (yes, we know it’s stupid, blame Google).
> 
> If you have questions, concerns, or need clarification, please reply to this email and just ask us.
> 
> (While we have tried to make this review as exhaustive as possible we, like you, are humans and may have missed things. As such, we will re-review the ENTIRE plugin when you send it back to us. > We appreciate your patience and understanding in this.)

**Comment by jonathanstegall** [comment]
Things I've done:

1. Rename all classes and declarations (looked through every time we say `new ___`.
2. Rename all defined constants.
3. Rename the places where we call the classes by name.
4. Look through all cases of `do_action` (they seem fine already) and `apply_filters` (they seem fine also).
5. Look through all cases of `update_option` (they seem fine already). We have no cases of `add_option`.
6. Edit/regenerate docs.

**Comment by jonathanstegall** [comment]
Admin.php seems to be the only place where we need to sanitize or validate. Escape is maybe more complicated though.

**Comment by benlk** [comment]
Yep, because the only place that's not `classes/admin.php` is in `classes/salesforce_pull.php` and it's only checking for `$_POST`'s non-emptiness: https://github.com/MinnPost/object-sync-for-salesforce/blob/master/classes/salesforce_pull.php#L94-L98

Shall I run through and sanitize all the `$_POST` data?

As far as escaping goes, what are your concerns there? There will be stuff to escape in:
- the admin
- the logging class and its logs
- the content and fields of posts created by the sync actions
- ?

**Comment by jonathanstegall** [comment]
@benlk sure, feel free. My only concern is that I don't really know which escape functions to use, when they might make things different than I expect, when I might miss items, etc.

**Comment by jonathanstegall** [comment]
@benlk But maybe we should merge #86 first. @rclations, do you think it's good?

**Comment by benlk** [comment]
#88 merged, taking care of #86. I'll start on the sanitizing and validating.

**Comment by jonathanstegall** [comment]
I've made a branch for this issue if you want to use it, @benlk. But feel free to use another if you've got it. Whatever works. https://github.com/MinnPost/object-sync-for-salesforce/tree/87-fix-plugin-review-issues-2

**Comment by jonathanstegall** [comment]
Fixed in #93
